### PR TITLE
Tcp message type id fix

### PR DESCRIPTION
**Pull request by condron**
Cleanup on the tcp type identification on deserialization to handle known issue with message hierarchies with different message sets. If both sides do not have identical message hierarchies the type identifier is not guaranteed to the same.

Added serialization tests
Added full type name information dictionary to `MessgeHierarchy`
Clarified intent on  the `tcpmessage` struct 
Removed double serialization on generating `tcpmessage` from byte array
added support for multiple clients

### CODE REVIEW

File: src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs [hunk 1]
```diff
@@ -35,17 +42,29 @@ public static TcpMessage FromArraySegment(ArraySegment<byte> data)
...
             }
             Log.Debug("Deserialized Message MsgId=" + msg.MsgId + " MsgTypeId=" + msg.MsgTypeId);
-            return new TcpMessage(msg);
+            return new TcpMessage(msg, data);
         }
-
+        //used by FromFramedArraySegment to set the values and return the struct
+        private TcpMessage(Message message, ArraySegment<byte> data)
+        {
+            MessageId = message.MsgTypeId;
```
Comment (StephenDrewLDS) [inline src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: How do you ensure that the message hierarchies on both sides of the TCP connection are identical?

Comment (condron) [reply src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: You don't. So the lookup by full type name in allows the receiving system to find the correct related type and populate is using weak serialization (giving us cross version support) and if the type isn't found the tcp connector logs a warning and moves on. 

We may want to add something like a dead letter queue to the workstation host, and consider how to propagate messages with unknown types in its the distribution logic. But that needs to be handled one layer up from here.

Comment (StephenDrewLDS) [reply src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: So why serialize it at all then?

Comment (condron) [reply src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: to be able to packetize it and send over tcp-ip efficiently? and if we don't serialize it then we need matching types on both sides.
also looking at the code displayed by GitHub in the review reminds me why we want upsource.
```
 using (var reader = new BsonDataReader(ms))
            {
                reader.Read(); //object
                reader.Read(); //property name

                var messageType = MessageHierarchy.GetMsgType((string)reader.Value);
                reader.Read(); //property value
                msg = (Message)JsonConvert.DeserializeObject((string)reader.Value, messageType);
            }
            Log.Debug("Deserialized Message MsgId=" + msg.MsgId + " MsgTypeId=" + msg.MsgTypeId);
            return new TcpMessage(msg, data);
```

Comment (condron) [reply src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: should I add a cross version  example to the demo?

Comment (StephenDrewLDS) [reply src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: It's slightly confusing that `MessageId` is actually `MsgTypeId`.... and I'm not clear why this needs to be serialized at all if it's not used to deserialize the message... shouldn't it be `JsonIgnore`d in that case and just used locally in-process?

Comment (condron) [reply src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: MsgId and MsgTypeId are different, and the MsgTypeId has a the json ignore attribute on it, so it is not serialized.
```
 public abstract class Message
    {
        protected static int NextMsgId = -1;
        private static readonly int TypeId = Interlocked.Increment(ref NextMsgId);
        [JsonIgnore]
        public virtual int MsgTypeId => TypeId;

        [JsonProperty(Required = Required.Always)]
        public Guid MsgId { get; private set; }

        protected Message()
        {
            MsgId = Guid.NewGuid();
        }
    }
```

Comment (StephenDrewLDS) [reply src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: Sure. But...

```csharp
MessageId = message.MsgTypeId;
```

Comment (condron) [reply src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: just added the core json settings to the serializer
```
public TcpMessage(Message message)
        {
            Ensure.NotNull(message, nameof(message));
            MessageId = message.MsgTypeId;
            MessageType = message.GetType();
            Log.Debug("Message MsgId=" + message.MsgId + " MsgTypeId=" + message.MsgTypeId + " to be wrapped.");

            var ms = new MemoryStream();
            using (var writer = new BsonDataWriter(ms))
            {
                writer.WriteStartObject();
                writer.WritePropertyName(MessageType.FullName);
                writer.WriteValue(JsonConvert.SerializeObject(message, Settings.JsonSettings));
                writer.WriteEnd();
            }
            Data = new ArraySegment<byte>(ms.ToArray());
            WrappedMessage = message;
        }
```

Comment (condron) [reply src/ReactiveDomain.Transport/CommandSocket/TcpMessage.cs]: ah, I missed that he had done that. I thought it was something else. (message length)

**Pull request merged**